core/src/main/scala/org/http4s/HeaderV2.scala

```diff
@@ -0,0 +1,292 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.header.v2
+
+import cats.{Id, Monoid, Order, Show}
+import cats.data.NonEmptyList
+import cats.syntax.all._
+import org.typelevel.ci.CIString
+import scala.collection.mutable.{ListBuffer, Set => MutSet}
+
+/** Typeclass representing an HTTP header, which all the http4s
+  * default headers satisfy.
+  * You can add custom headers by providing an implicit instance of
+  * `Header[YourCustomHeader]`
+  */
+trait Header[A, T <: Header.Type] {
+
+  /** Name of the header. Not case sensitive.
+    */
+  def name: CIString
+
+  /** Value of the header, which is represented as a String.
+    * Will be a comma separated String for headers with multiple values.
+    */
+  def value(a: A): String
+
+  /** Parses the header from its String representation.
+    * Could be a comma separated String in case of a Header with
+    * multiple values.
+    */
+  def parse(headerValue: String): Option[A]
+}
+object Header {
+  case class Raw(name: CIString, value: String, recurring: Boolean = true)
+  object Raw {
+    implicit val showForRawHeaders: Show[Raw] =
+      h => s"${h.name.show}: ${h.value}"
+
+    implicit lazy val orderForRawHeaders: Order[Raw] =
+      Order.whenEqual(
+        Order.by(_.name),
+        Order.by(_.value)
+      )
+  }
+
+  /** Classifies custom headers into `Single` headers, which can only
+    * appear once, and `Recurring` headers, which can appear multiple
+    * times.
+    */
+  sealed trait Type
+  case class Single() extends Type
+  case class Recurring() extends Type
+
+  def apply[A](implicit ev: Header[A, _]): ev.type = ev
+
+  /** Target for implicit conversions to Header.Raw from custom
+    * headers and key-value pairs.
+    *
+    * A method taking variadic `ToRaw` arguments will allow taking
+    * heteregenous arguments, provided they are either:
+    *
+    * - A value of type `A`  which has a `Header[A]` in scope
+    * - A (name, value) pair of `String`, which is treated as a `Recurring` header
+    * - A `Header.Raw`
+    *
+    * @see [[org.http4s.Headers$.apply]]
+    */
+  sealed trait ToRaw {
+    def value: Header.Raw
+  }
+  object ToRaw {
+    implicit def rawToRaw(h: Header.Raw): Header.ToRaw = new Header.ToRaw {
+      val value = h
+    }
+
+    implicit def keyValuesToRaw(kv: (String, String)): Header.ToRaw = new Header.ToRaw {
+      val value = Header.Raw(CIString(kv._1), kv._2)
+    }
+
+    implicit def customHeadersToRaw[H](h: H)(implicit H: Header[H, _]): Header.ToRaw =
```

Comment:
`custom` to me sounds like a third-party addition.  akka-http calls these `"modeled"` headers, which I like.

It doesn't matter too much, since it's implicit.

Reply:
It makes sense, I've used "custom" in the scaladoc too